### PR TITLE
Preserve density - see #30

### PR DESCRIPTION
**Pull request by lopuhin**
By default density is preserved (new behaviour) for char and char_wb analyzers, and old behaviour is kept for word analyzer, as discussed in #30.

### CODE REVIEW

**Comment by lopuhin** [comment]
Hm, I re-triggered the travis build by squashing commits, but the codecov is still showing low coverage of the patch.

**Comment by kmike** [comment]
Is it rebased on master?

**Comment by lopuhin** [comment]
It was rebased against an old master, I forgot to pull - rebased against up-to-date master now, thanks for the hint!

**Comment by lopuhin** [comment]
Hm, maybe codecov is correct after all :)

**Comment by lopuhin** [comment]
Yeah, it was really lack of tests on my part, fixed now.

File: eli5/formatters/html.py [hunk 1]
```diff
@@ -49,19 +51,29 @@ def format_html_styles():
...
     for _, spans, weight in weighted_spans:
         for start, end in spans:
+            if preserve_density:
+                weight /= (end - start)
             char_weights[start:end] += weight
     # TODO - can be much smarter, join spans at least
     # TODO - for longer documents, remove text without active features
     not_found_weights = sorted(
-        (feature, weight)
+        (feature, weight / len(feature) if preserve_density else weight)
```
Comment (kmike) [inline eli5/formatters/html.py]: This part is tricky - non-text feature names can be lighter or darker depending on a length of their name; two features with equal weights can have different color if their names have different length. I don't know how to solve it :) My main motivation was to make highlighting compatible between char and token models, not so much between text and non-text features.

Maybe it deserves its own solution - e.g. instead of showing non-text features inline before the text we may show them separated by new lines, and near each feature name show a rectangle, all rectangles of the same size. This way length of non-text feature name won't impact highlighting. To make it compatible with text rectangle may have a width similar to avg. word width (e.g. 5 chars).

**Comment by lopuhin** [comment]
Another thing that does not play nicely with the idea of preserving density is that right now, we take the feature weight that includes the vectorizer score. As a result, if a feature is repeated several times, then its weight might be higher (if using CountVectorizer for example), but it will be highlighted with this higher weight several times. So on the screenshot "space" is more important feature, which is obvious from the first example, but not from the second, where "flight" is repeated several times:


So I think we should evenly divide the weight of a feature among all it's instances, and do it for all analyzers (for words and chars). Even with CountVectorizer(binary=True) that will give the correct impression I think.

File: eli5/formatters/html.py (hunk 1 above)
Comment (lopuhin) [inline eli5/formatters/html.py]: > but unfortunately our text highlighting already works only for linear models
Because of adding weights for overlapping features

File: eli5/formatters/features.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+
+
+class EscapedFeatureName(object):
```
Comment (kmike) [inline eli5/formatters/features.py]: What do you think about renaming it to something like FormattedFeatureName or PreparedFeatureName? If I understood the code correctly, instances of this class are passed as-is through formatting functions, this is not directly related to HTML escaping.

Comment (lopuhin) [inline eli5/formatters/features.py]: Yes, I like `FormattedFeatureName`, thanks for the suggestion! Changed it.

File: eli5/formatters/features.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    """ Feature name that does not need any additional escaping.
+    """
+    def __init__(self, value):
+        self.value = value
+
+    def __str__(self):
+        return str(self.value)
```
Comment (lopuhin) [inline eli5/formatters/features.py]: Right, I didn't think about Python 2 here, thanks for the catch!
I want it to return text type, so it seems easier to use a `.format()` method instead: ee0c4ae.

File: tests/test_formatters_text.py [hunk 4]
```diff
@@ -1,11 +1,11 @@
 # -*- coding: utf-8 -*-
 
-from eli5.formatters import EscapedFeatureName
+from eli5.formatters import FormattedFeatureName
 from eli5.formatters.text import _format_feature, _SPACE
 
 
 def test_format_escaped_feature():
```
Comment (kmike) [inline tests/test_formatters_text.py]: it makes sense to change test names as well

Comment (lopuhin) [inline tests/test_formatters_text.py]: Right, I missed that, thanks! Added to 1a98be8

File: eli5/formatters/features.py [hunk 5]
```diff
@@ -1,14 +1,16 @@
...
 class FormattedFeatureName(object):
     """ Feature name that does not need any additional escaping.
     """
     def __init__(self, value):
         self.value = value
 
-    def __str__(self):
-        return str(self.value)
+    def format(self):
+        return six.text_type(self.value)
```
Comment (kmike) [inline eli5/formatters/features.py]: What's six.text_type call for? When self.value can be bytes? In Python 2.x `unicode(b'string')` works correctly only for ascii-only byte strings because it tries to decode bytes data using sys.getdefaultencoding(). In Python3 str(b'foo') returns `"b'foo'"`, not `"foo"`.

Comment (lopuhin) [inline eli5/formatters/features.py]: Fixed now. Didn't think about python 2 string for a while, sorry for the mess :)
